scripts/check-components

```diff
@@ -0,0 +1,258 @@
+# Transforms implicitly handle the common events in the topology, and
+# have no other emit requirements.
+TRANSFORM_EVENTS = []
+
+$error_count = 0
+
+def hash_array_add(hash, key, item)
+  arr = hash.fetch(key, Array::new)
+  arr.append(item)
+  hash[key] = arr
+end
+
+# Reduce a path to a component name, used to group all the emits for a
+# component that may be split over multiple files.
+def path_component(path)
+  parts = path.split('/')
+  parts.shift # Drop the leading 'src'
+  group = parts.shift
+  component = parts.shift
+
+  # Skip the top-level module files and utilities
+  if component == 'mod.rs' or component == 'util'
+    return nil
+  end
+
+  if component == 'gcp'
+    component = "gcp_#{parts.shift}"
+  end
+
+  if component.end_with? '.rs'
+    component = component[..-4]
+  end
+  return "#{group}/#{component}"
+end
+
+# A class to hold error reports and common functionality
+class Event
+  def initialize(name)
+    @path = nil
+    @name = name
+    @reports = []
+    @members = []
+    @counters = []
+    @logs = []
+  end
+
+  def path(path)
+    @path = path
+  end
+
+  def members(list)
+    @members = list
+  end
+
+  def add_counter(name, tags)
+    @counters.append(name)
+  end
+
+  def add_log(type, message, parameters)
+    @logs.append([type, message, parameters])
+  end
+
+  def validate
+    # Check BytesReceived events (for sources)
+    if @name.end_with? 'BytesReceived'
+      members_must_include(['byte_size'])
+      counters_must_include(['received_bytes_total'])
+    end
+
+    # Check EventsReceived events (common)
+    if @name.end_with? 'EventsReceived'
+      members_must_include(['count', 'byte_size'])
+      counters_must_include(['received_events_total', 'received_event_bytes_total'])
+    end
+
+    # Check EventsSent events (common)
+    if @name.end_with? 'EventsSent'
+      members_must_include(['count', 'bytes_sent'])
+      counters_must_include(['sent_events_total', 'sent_event_bytes_total'])
+    end
+
+    # Check BytesSent events (for sinks)
+    if @name.end_with? 'BytesSent'
+      members_must_include(['byte_size'])
+    end
+
+    has_errors = @logs.one? { |type, _, _| type == 'error' }
+
+    # Make sure Error events output an error
+    if has_errors or @name.end_with? 'Error'
+      append('Error events MUST be named "___Error".') unless @name.end_with? 'Error'
+      counters_must_include(['errors_total'])
+    end
+
+    # Make sure error events contain the right parameters
+    @logs.each { |type, message, parameters|
+      if type == 'error'
+        ['error', 'stage'].each { |parameter|
+          unless parameters.include? parameter
+            @reports.append("Error log MUST include parameter \"#{parameter}\".")
+          end
+        }
+      end
+    }
+
+    unless @reports.empty?
+      puts "#{@path}: Errors in event #{@name}:"
+      @reports.each { |report|
+        puts "    #{report}"
+        $error_count += 1
+      }
+    end
```

Comment:
Using `puts` is bad form within a method call. It should be done from the caller's perspective. Something like:

```ruby
event = Event.new
if !event.valid?
  event.reports.each { |r| puts r }
end
```